pkg/dns/treecache.go

```diff
@@ -0,0 +1,312 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dns
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"reflect"
+	"strings"
+	"sync"
+)
+
+const (
+	dataFile = "data.dat"
+	crcFile  = "data.crc"
+)
+
+type object interface{}
+
+type TreeCache struct {
+	ChildNodes map[string]*TreeCache
+	Entries    map[string]interface{}
+	m          *sync.RWMutex
+}
+
+func NewTreeCache() *TreeCache {
+	return &TreeCache{
+		ChildNodes: make(map[string]*TreeCache),
+		Entries:    make(map[string]interface{}),
+		m:          &sync.RWMutex{},
+	}
+}
+
+func Deserialize(dir string) (*TreeCache, error) {
+	b, err := ioutil.ReadFile(path.Join(dir, dataFile))
+	if err != nil {
+		return nil, err
+	}
+	var hash []byte
+	hash, err = ioutil.ReadFile(path.Join(dir, crcFile))
+	if err != nil {
+		return nil, err
+	}
+	if !reflect.DeepEqual(hash, getMD5(b)) {
+		return nil, fmt.Errorf("Checksum failed")
+	}
+
+	var cache TreeCache
+	err = json.Unmarshal(b, &cache)
+	if err != nil {
+		return nil, err
+	}
+	cache.m = &sync.RWMutex{}
+	return &cache, nil
+}
+
+func (cache *TreeCache) Serialize(dir string) (string, error) {
+	cache.m.RLock()
+	defer cache.m.RUnlock()
+	b, err := json.Marshal(cache)
+	if err != nil {
+		return "", err
+	}
+
+	if len(dir) == 0 {
+		var prettyJSON bytes.Buffer
+		err = json.Indent(&prettyJSON, b, "", "\t")
+
+		if err != nil {
+			return "", err
+		}
+		return string(prettyJSON.Bytes()), nil
+	}
+	if err := ensureDir(dir, os.FileMode(0755)); err != nil {
+		return "", err
+	}
+	if err := ioutil.WriteFile(path.Join(dir, dataFile), b, 0644); err != nil {
+		return "", err
+	}
+	if err := ioutil.WriteFile(path.Join(dir, crcFile), getMD5(b), 0644); err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+func (cache *TreeCache) SetEntry(key string, val interface{}, path ...string) {
+	cache.m.Lock()
+	defer cache.m.Unlock()
+	node := cache.ensureChildNode(path...)
+	node.Entries[key] = val
+}
+
+func (cache *TreeCache) ReplaceEntries(entries map[string]interface{}, path ...string) {
+	cache.m.Lock()
+	defer cache.m.Unlock()
+	node := cache.ensureChildNode(path...)
+	node.Entries = make(map[string]interface{})
+	for key, val := range entries {
+		node.Entries[key] = val
+	}
+}
+
+func (cache *TreeCache) GetSubCache(path ...string) *TreeCache {
```

Comment:
Discussed offline with Tim. locking isnt busted..
Made a modification so that subcaches have nil mutex value, since those mutexes are not used any way.